agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/TelemetryClient.java

```diff
@@ -245,20 +237,26 @@ public BatchSpanProcessor getStatsbeatChannelBatcher() {
     if (statsbeatChannelBatcher == null) {
       synchronized (channelInitLock) {
         if (statsbeatChannelBatcher == null) {
-          LocalFileCache localFileCache = new LocalFileCache();
-          File statsbeatFolder = LocalStorageUtils.getOfflineStatsbeatFolder();
-          LocalFileLoader localFileLoader =
-              new LocalFileLoader(localFileCache, statsbeatFolder, null);
-          LocalFileWriter localFileWriter =
-              new LocalFileWriter(localFileCache, statsbeatFolder, null);
+          LocalFileLoader localFileLoader = null;
+          LocalFileWriter localFileWriter = null;
+          if (!readonly) {
+            LocalFileCache localFileCache = new LocalFileCache();
+            File statsbeatFolder = LocalStorageUtils.getOfflineStatsbeatFolder();
+            localFileLoader = new LocalFileLoader(localFileCache, statsbeatFolder, null);
+            localFileWriter = new LocalFileWriter(localFileCache, statsbeatFolder, null);
+          }
+
           TelemetryChannel channel =
               TelemetryChannel.create(
                   endpointProvider.getStatsbeatEndpointUrl(),
                   localFileWriter,
                   ikeyEndpointMap,
                   null,
                   null);
-          LocalFileSender.start(localFileLoader, channel);
+
+          if (!readonly) {
+            LocalFileSender.start(localFileLoader, channel);
+          }
```

Comment:
can this be moved up into the above `readonly` condition?

Reply:
no, needs to be after channel is defined.